packages/nimble-components/src/rich-text-mention/base/models/mention-validator.ts

```diff
@@ -0,0 +1,109 @@
+import type { Mapping } from '../../../mapping/base';
+import { Validator, ValidityObject } from '../../../utilities/models/validator';
+import type { MentionInternals } from './mention-internals';
+
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface RichTextMentionValidity extends ValidityObject {}
+
+type FlagNames<T> = T extends readonly (infer U)[] ? U : never;
+
+export const baseValidityFlagNames = [
+    'unsupportedMappingType',
+    'duplicateMappingMentionHref',
+    'missingMentionHrefValue',
+    'unsupportedMentionHrefValue',
+    'missingPatternAttribute'
+] as const;
+
+/**
+ * Validator for RichTextMention
+ */
+export class RichTextMentionValidator<
+    ValidityFlagNames extends readonly string[]
+> extends Validator<typeof baseValidityFlagNames | ValidityFlagNames> {
+    public constructor(
+        private readonly mentionInternals: MentionInternals<unknown>,
+        configValidityKeys: ValidityFlagNames,
+        private readonly supportedMappingElements: readonly (typeof Mapping<unknown>)[]
+    ) {
+        super(configValidityKeys);
+    }
+
+    public validate(
+        mappings: Mapping<unknown>[],
+        pattern: string | undefined
+    ): void {
+        this.untrackAll();
+        const mentionHrefs = mappings.map(mapping => mapping.key);
+        this.validateMappingTypes(mappings);
+        this.validateNoMissingMentionHref(mentionHrefs);
+        this.validateUniqueMentionHref(mentionHrefs);
+        this.validatePattern(pattern);
+        this.validateHref(mentionHrefs, pattern);
+    }
+
+    /**
+     * @returns an object containing flags for various ways the configuation can be invalid
+     */
+    public getValidity(): RichTextMentionValidity {
+        return this.getValidationFlags();
+    }
+
+    /**
+     * Sets a particular validity condition flag's value, e.g. "hasInvalidFooValue" = true
+     */
+    protected setConditionValue(
+        name: FlagNames<typeof baseValidityFlagNames | ValidityFlagNames>,
+        isInvalid: boolean
+    ): void {
+        if (isInvalid) {
+            this.track(name);
+        } else {
+            this.untrack(name);
+        }
+        this.updateMentionInternalsFlag();
+    }
+
+    private updateMentionInternalsFlag(): void {
+        this.mentionInternals.validConfiguration = this.isValid();
+    }
+
+    private validatePattern(pattern: string | undefined): void {
```

Comment:
pattern should be tested against the RegExp constructor which can throw: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/RegExp#exceptions